llvm/lib/Object/SYCLBIN.cpp

```diff
@@ -0,0 +1,400 @@
+    if (!ReadSizeOrError)
+      return ReadSizeOrError.takeError();
+    if (Error EC = ErrorIfSizeUnavailable(*ReadSizeOrError))
+      return EC;
+    return *ReadSizeOrError;
+  }
+
+  template <typename ReadSizePromiseT>
+  Expected<llvm::StringRef> ConsumeStringRef() {
+    Expected<ReadSizePromiseT> StringSizeOrError =
+        ConsumeReadSizePromise<ReadSizePromiseT>();
+    if (!StringSizeOrError)
+      return StringSizeOrError.takeError();
+    llvm::StringRef Result{GetCurrentPointer(), size_t{*StringSizeOrError}};
+    Move(*StringSizeOrError);
+    return Result;
+  }
+
+  template <typename ReadSizePromiseT>
+  Expected<SmallString<0>> ConsumeString() {
+    Expected<StringRef> StringRefOrError = ConsumeStringRef<ReadSizePromiseT>();
+    if (!StringRefOrError)
+      return StringRefOrError.takeError();
+    return static_cast<SmallString<0>>(*StringRefOrError);
+  }
+
+  Expected<SmallVector<SmallString<0>>> ConsumeStringList() {
+    ConsumerParser ListConsumer;
+    if (Error EC = CreateSubConsumer<uint64_t>().moveInto(ListConsumer))
+      return EC;
+
+    Expected<uint32_t> NumStringsOrError =
+        ListConsumer.ConsumeScalar<uint32_t>();
+    if (!NumStringsOrError)
+      return NumStringsOrError.takeError();
+
+    SmallVector<SmallString<0>> Result;
+    Result.reserve(*NumStringsOrError);
+    for (size_t I = 0; I < *NumStringsOrError; ++I) {
+      Expected<SmallString<0>> StringOrError =
+          ListConsumer.ConsumeString<uint32_t>();
+      if (!StringOrError)
+        return StringOrError.takeError();
+      Result.emplace_back(*StringOrError);
+    }
+
+    return Result;
+  }
+
+  size_t GetRemainingSize() const noexcept { return RemainingSize; }
+
+  bool Empty() const noexcept { return GetRemainingSize() == 0; }
+
+private:
+  Error ErrorIfSizeUnavailable(size_t Size) const {
+    if (RemainingSize < Size)
+      return createStringError(inconvertibleErrorCode(),
+                               "Incorrect SYCLBIN magic number.");
+    return Error::success();
+  }
+
+  const char *GetCurrentPointer() const noexcept { return Data; }
+
+  void Move(size_t Size) noexcept {
+    Data += Size;
+    RemainingSize -= Size;
+  }
+
+  const char *Data = nullptr;
+  size_t RemainingSize = 0;
+};
+
+} // namespace
+
+Expected<SmallString<0>>
+SYCLBIN::write(const SmallVector<SYCLBIN::ModuleDesc> &ModuleDescs) {
+  // TODO: Merge by properties and kernel names, so overlap can live in the same
+  //       abstract module.
+
+  SmallString<0> Data;
+  raw_svector_ostream OS(Data);
+  OS << StringRef(reinterpret_cast<const char *>(&MagicNumber),
+                  sizeof(MagicNumber));
+  BinaryWriteInteger<uint32_t>(OS, Version);
+  BinaryWriteInteger<uint8_t>(OS, 0); // TODO: Deduce this from arguments.
+
+  {
+    SmallString<0> BodyData;
+    raw_svector_ostream BodyOS(BodyData);
+    for (const ModuleDesc &Desc : ModuleDescs) {
+      for (const module_split::SplitModule &SM : Desc.SplitModules) {
```

Comment:
This is a very long `for` loop, I suggest that we outline some elements of it into smaller helper functions so that it is easier to grasp high-level logic and keep track of indentation levels

Reply:
I've separated it a bit.